### PR TITLE
Allow for a kind of equality for decoders

### PR DESCRIPTION
**Pull request by rgrempel**
In Elm, a JSON Decoder is, from one point of view, simply a function which takes a `Value` as an input, and produces something as an output. In fact, in implementing `Elm.Json.Decode`, I have been defining it in this way:

    newtype Decoder a = Decoder (Value -> Result String a)

... and it's perfectly possible to implement the Elm decoder API on this basis.

However, that's not how Elm actually implements a decoder -- it creates a kind of DSL, collecting information about how to proceed, and only actually "running" the DSL when needed.

My theory is that Elm does that because it needs to test the equality of decoders -- it has an `equalDecoders` function for this purpose. I believe the reason Elm wants to test the equality of decoders is to be able to test the equality of HTML nodes in the virtual DOM. A decoder is an essential part of various event handlers. Thus, to test whether an event handler has changed, and thus a listener needs to be removed and/or applied, you need to test the equality of the decoder you got in the previous round with the decoder you got in this round. Unless you can determine that the decoders are equal, you'll need to remove the previous listener and add it again, which would be inefficient (at the very least).

Now, if a `Decoder` were just a `Function`, there would be very little you could do to test for equality.

- You can use referential equality, but that is subject to false negatives ... two functions can be equal without being the very same function. (E.g. lambdas that have the same inputs, but are produced on separate occasions).

- Some Javascript implementations will give you the source code of the function as a string if you ask. However, I don't think this is actually reliable as an equality test because you wouldn't necessarily have closed over the same values. (I don't think this affects the referential equality check, since you couldn't have closed over different values if you're referentially equal).

- In theory, you could call the functions on the entire input domain and compare the outputs, but that's only practical in specialized situations. (It would be practical in some situations where the input domain isn't very large).

So, basically you're stuck with referential equality ... the two decoders would be equal if they were the very same function.

What I think the Elm implementation is trying to do is make judging equality slightly more accurate, especially when composing decoders together. If you compose two functions together, the result will not be referentially equal unless you hang onto a reference and use it. However, in the Elm DSL, the composition of two decoders doesn't "forget" the original decoders. So, if the inputs are referentially equal, it's still possible to determine that the composed decoders are equal, even if you don't retain a reference to the composed decoder.

So, what I'm attempting to do on this PR is to be at least as good as Elm is when judging the equality of decoders.

### CODE REVIEW

**Comment by rgrempel** [comment]
So far, I've got a DSL which passes the original tests. The next step is to actually write the `equalDecoders` function, and then adjust the DSL to make it as reliable as possible. (Where "reliable" means avoiding as many false negatives as possible ... we won't tolerate any false positives).

I won't make an `Eq` instance for `Decoder`, since it would not really be lawful ... it's a kind of ersatz equality.

I had been working on a somewhat more ambitious `EqFunc` idea, where ordinary functions could be lifted into an `EqFunc`, which could then be manipulated in various equality-preserving ways. That turned out to be a little too ambitious, but provided some of the ideas I'm using here.

I'll plan to end up documenting how to preserve good behaviour for `equalDecoders` when constructing decoders -- this would actually be useful in Elm itself as well, since the same rules of thumb would apply, and there isn't anything else telling you whether you are creating decoders that will successfully test for equality.

**Comment by rgrempel** [comment]
There are still some imperfections here ... notably, `keyValuePairs` and `dict` should be able to work with `equalDecoders`, and I should try to do something about making the `Apply` instance work better with `equalDecoders`. However, it's good enough for now, so I'll merge it and move on!
